reactor-core/src/main/java21/reactor/core/scheduler/LoomBoundedElasticScheduler.java

```diff
@@ -0,0 +1,1352 @@
+				if (previous.shutdown) {
+					return false;
+				}
+
+				int len = previous.array.length;
+				BoundedState[] replacement = new BoundedState[len + 1];
+				System.arraycopy(previous.array, 0, replacement, 0, len);
+				replacement[len] = bs;
+
+				if (BUSY_STATES.compareAndSet(this, previous,
+						new BusyStates(replacement, false))) {
+					return true;
+				}
+			}
+		}
+
+		void remove(BoundedState boundedState) {
+			for(;;) {
+				BusyStates current = busyStates;
+				BoundedState[] arr = busyStates.array;
+				int len = arr.length;
+
+				if (len == 0 || current.shutdown) {
+					return;
+				}
+
+				BusyStates replacement = null;
+				if (len == 1) {
+					if (arr[0] == boundedState) {
+						replacement = ALL_IDLE;
+					}
+				}
+				else {
+					for (int i = 0; i < len; i++) {
+						BoundedState state = arr[i];
+						if (state == boundedState) {
+							replacement = new BusyStates(
+									new BoundedState[len - 1], false
+							);
+							System.arraycopy(arr, 0, replacement.array, 0, i);
+							System.arraycopy(arr, i + 1, replacement.array, i, len - i - 1);
+							break;
+						}
+					}
+				}
+				if (replacement == null) {
+					//bounded state not found, ignore
+					return;
+				}
+				if (BUSY_STATES.compareAndSet(this, current, replacement)) {
+					return;
+				}
+			}
+		}
+
+		/**
+		 * Pick a {@link BoundedState}, prioritizing idle ones then spinning up a new one if enough capacity.
+		 * Otherwise, picks an active one by taking from a {@link PriorityQueue}. The picking is
+		 * optimistically re-attempted if the picked slot cannot be marked as picked.
+		 *
+		 * @return the picked {@link BoundedState}
+		 */
+		BoundedState pick() {
+			for (;;) {
+				if (busyStates == ALL_SHUTDOWN) {
+					return CREATING; //synonym for shutdown, since the underlying executor is shut down
+				}
+
+				int a = get();
+				if (a < parent.maxThreads) {
+					//try to build a new resource
+					if (compareAndSet(a, a + 1)) {
+						BoundedState newState = new BoundedState(this);
+						if (newState.markPicked()) {
+							boolean accepted = add(newState);
+							if (!accepted) { // shutdown in the meantime
+								newState.shutdown(true);
+								return CREATING;
+							}
+							return newState;
+						}
+					}
+					//else optimistically retry (implicit continue here)
+				}
+				else {
+					BoundedState s = choseOneBusy();
+					if (s != null && s.markPicked()) {
+						return s;
+					}
+					//else optimistically retry (implicit continue here)
+				}
+			}
+		}
+
+		@Nullable
+		private BoundedState choseOneBusy() {
+			BoundedState[] arr = busyStates.array;
+			int len = arr.length;
+			if (len == 0) {
+				return null; //implicit retry in the pick() loop
+			}
+			if (len == 1) {
+				return arr[0];
+			}
+
+			BoundedState choice = arr[0];
+			int leastBusy = Integer.MAX_VALUE;
+
+			for (int i = 0; i < arr.length; i++) {
+				BoundedState state = arr[i];
+				int busy = state.markCount;
```

Comment:
Consider using pending tasks queue size instead. Probably in a lot of cases the number of users of a boundedState will be similar, but the queue size is our current best tool to measure how busy the boundedState is.

Reply:
lets review that separately as an effort for both BoundedElasticSchedulers